numcodecs/compat.py

```diff
@@ -14,29 +14,48 @@
 
 if PY2:  # pragma: py3 no cover
 
+    buffer = buffer
     text_type = unicode
     binary_type = str
     integer_types = (int, long)
     reduce = reduce
 
 else:  # pragma: py2 no cover
 
+    buffer = memoryview
     text_type = str
     binary_type = bytes
     integer_types = int,
     from functools import reduce
 
 
+def to_buffer(v):
+    """Obtain a `buffer` or `memoryview` for `v`."""
+    b = v
+
+    if isinstance(b, np.ndarray):
+        if b.dtype.kind in 'Mm':
+            b = b.view(np.uint64)
+
+    try:
+        b = memoryview(b)
+    except TypeError:  # pragma: py3 no cover
+        pass
+    else:
+        if b.format is not 'O':
+            b = np.array(b, copy=False).reshape(-1, order='A')
```

Comment:
I think the intention here is to ensure the buffer is read-only, but I'm afraid doing it this way will introduce some unintended memory copies. You could use np.frombuffer() instead which should guarantee no memory copy, but then you'd have to provide a dtype IIRC, and I think we want to interfere as little as possible with type information being passed through the buffer interface. Hence I was wondering if we can just do without this, i.e., pass the buffer/memoryview through and don't worry about whether it's read only or not.

Reply:
Actually `np.array(b, copy=False)` does not copy `memoryview`s or `buffer`s. It takes a view. Have verified this in experiments on both Python 2 and 3. This has no effect on whether the buffer is readable or not (though it will inherit whatever setting the underlying buffer has).

Reply:
Should add the main goal of this operation is to ensure that the data is flattened since many compressors expect this (and were doing that before).

Reply:
Should note have included a test using NumPy's [`shares_memory`]( https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.shares_memory.html ) to demonstrate this, which hopefully alleviates this concern.